OpenRA.Mods.Common/AI/HackyAI.cs

```diff
@@ -250,6 +270,60 @@ public void Activate(Player p)
 				resourceTypeIndices.Set(World.TileSet.GetTerrainIndex(t.TerrainType), true);
 		}
 
+		// TODO: Possibly give this a more generic name when terrain type is unhardcoded
+		public bool EnoughWaterToBuildNaval()
+		{
+			var baseProviders = World.Actors.Where(
+				a => a.Owner == Player
+					&& a.HasTrait<BaseProvider>()
+					&& !a.HasTrait<Mobile>());
+
+			foreach (var b in baseProviders)
+			{
+				// TODO: Unhardcode terrain type
+				// TODO2: Properly check building foundation rather than 3x3 area
+				var playerWorld = Player.World;
+				var countWaterCells = Map.FindTilesInCircle(b.Location, Info.MaxBaseRadius)
+					.Where(c => playerWorld.Map.Contains(c)
+						&& playerWorld.Map.GetTerrainInfo(c).IsWater
+						&& Util.AdjacentCells(playerWorld, Target.FromCell(playerWorld, c))
+							.All(a => playerWorld.Map.GetTerrainInfo(a).IsWater))
+					.Count();
+
+				if (countWaterCells > 0)
+					return true;
+			}
+
+			return false;
+		}
+
+		// Check whether we have at least one building providing buildable area close enough to water to build naval structures
+		public bool CloseEnoughToWater()
+		{
+			var areaProviders = World.Actors.Where(
+				a => a.Owner == Player
+					&& a.HasTrait<GivesBuildableArea>()
+					&& !a.HasTrait<Mobile>());
+
+			foreach (var a in areaProviders)
+			{
+				// TODO: Unhardcode terrain type
```

Comment:
No, the idea is that in the future this whole check could be generalized to check for terrain type(s) of the modders choice, so the TODO still stands.